### ISSUE TITLE
Won't run on OS X El Capitan due to kext signing

### ISSUE THREAD

**Issue by hristost** [RogueAmoeba/Soundflower#39]
Soundflower won't run under the newest OS X:

```
Hristos-MacBook-Pro:~ Hristo$ sudo kextutil /System/Library/Extensions/Soundflower.kext
Diagnostics for /System/Library/Extensions/Soundflower.kext:
Code Signing Failure: not code signed
Untrusted kexts are not allowed
ERROR: invalid signature for com.Cycling74.driver.Soundflower, will not load
```

Is there a way we can get around this?

**Comment by LucaTNT** [comment]
Try enabling unsigned kext loading: `sudo nvram kext-dev-mode=1` and reboot.

**Comment by parvi** [comment]
Not working with me too. Just updated El Capitan was working fine in Yosemite. CAnt see it in applications too :( :(

**Comment by mattingalls** [comment]
try this version: https://github.com/mattingalls/Soundflower/releases/tag/2.0b2

**Comment by NVY-NUKULAR** [comment]
thx for the link! i now have my 2 channels back!

they seem to be working as well, but how do i get the soundflowerbed app to work again.

i am trying to stream to twitch using the OBS app. the soundflowerchannels are working and the OBS is getting the audio signal like before. but now i cant hear anything anymore, because im not able to tell soundflowerbed to output the channel signal. I was able to dicribe my problem so you can understand.

any help is much appreciated!

Nuku

**Comment by mattingalls** [comment]
unfortunately soundflowebed needs a lot of work. (better to just replace it, actually).  if all you need is to monitor soundflower channels, try going into Audio MIDI Setup and create a "Multi Output Device" with soundflower and your monitoring device.

**Comment by parvi** [comment]
@mattingalls I installed from the link you provided. It never show up in the applications (it was showing under applications and working perfectly under Yosemite) but in El Capitan havent found luck yet. Would you be working on improving this?? I simply install and try to run it using Spotlight but it could not be found. :( :( It is such a great app without which my external screen sound could not be controlled. Disappointed with this Mac update.

**Comment by zarroyorosa** [comment]
@mattingalls I love sounddlowebed and the fact that I can't use it like before disappoints me. I hope they update this app I really don't want to replace it.

**Comment by DopeyRunr** [comment]
@mattingalls Thanks so much for your version of Soundflower. It's working great. The only problem I'm having is that in the past I have modified the Info.plist file to add 12 more two-channel audio devices (sample attached below). I just cut and pasted that section, changing the bolded name of the channel each time to Soundflower - A, Soundflower - B etc. 
![screen shot 2015-10-02 at 8 09 10 am](https://cloud.githubusercontent.com/assets/14940164/10246199/f6e670bc-68dc-11e5-8430-e3fda0370680.png)
However, I am unable to do as I have in the past, which is to copy this Info.plist into the package, run these commands, and use all of those "virtual audio cables."

sudo chown root:wheel /Library/Extensions/Soundflower.kext/Contents/Info.plist
sudo chmod 755 /Library/Extensions/Soundflower.kext/Contents/Info.plist
sudo kextload /Library/Extensions/Soundflower.kext

I keep getting an error on the last step:

/Library/Extensions/Soundflower.kext failed to load - (libkern/kext) not loadable (reason unspecified); check the system/kernel logs for errors or try kextutil(8).

Any ideas on how to resolve this?

In case you're curious, I use all of the additional 2 channel devices like virtual audio cables to route output from several concurrent instances of Skype into separate tracks of Adobe Audition, create a sort of mix-minus and send combined audio from the other Skypes back to each of those Skype instances. Having all of those extra 2 channel devices means I can easily edit one participant's audio without changing other participants' audio.

Thanks so much for your help!

**Comment by evanedinger** [comment]
I just wanted to say thank you to @mattingalls for the link! I thought I'd be screwed for sure, but soundflower is now working just as usual! I made an account just to say this! Thanks!

**Comment by jeffreykog** [comment]
One way to fix these signing issues for good is to have a signed version of the kext like the one from @mattingalls which loads the sound device config from an external config file. That way we will be able to change the config file without breaking everything.

**Comment by verdantrama** [comment]
Since updating to El Capitan I was unable to use Soundflower with Screenflick. I uninstalled Soundflower and used @mattingalls solution. It worked perfectly and I am back in action. Like @evanedinger I made an account just to make this comment! Thanks!

**Comment by mattingalls** [comment]
@LucaTNT -- FYI i'm watching the WWDC session on security https://developer.apple.com/videos/play/wwdc2015-706/ and they say starting with El Capitan they have prevented disabled of signing with sudo nvram kext-dev-mode=1

**Comment by LucaTNT** [comment]
@mattingalls thanks for the heads up, I didn't catch that!

**Comment by worldsfinestdj** [comment]
If anyone is still experiencing this issue please go to the link I provide, it explains exactly how to have sound flower on the lastest Mac update El Captain https://youtu.be/kyS7UG3wmb0

**Comment by gilbe** [comment]
Awesome, thanks for the link @mattingalls worked a treat - Although still warned about unknown developer .. but installed and working now

**Comment by ManuelAbril** [comment]
Hi there, thank you so much for working so hard, @mattingalls. However, I have the following issues:
- I cannot see either he icon in the bar nor in applications.
- I followed these steps: uninstall previous version, reboot, installa lastest beta version, reboot. I swapped between 2ch and 64ch both sound and midi settings and I cannot listen anything.

**Comment by worldsfinestdj** [comment]
@manuelabril if you are still having trouble I posted a link above in the comments, I'm sure that would help you out.

**Comment by ManuelAbril** [comment]
@worldsfinestdj I saw it and I did everything you posted for soundflower installation because I do not know what it the purpose of the other app. Should I have to install?

**Comment by worldsfinestdj** [comment]
The other app is just what some people use to show gaming or basically record your screen , you do not need the other app, hopefully your sound flower is working now.

**Comment by ManuelAbril** [comment]
No it isn't. I usually launch the app and select 2ch or 64ch and it works. However, I cannot see the app anymore and I do not know how to open it.

**Comment by worldsfinestdj** [comment]
The app will not be visible itself but it's functions are still available if you go to system settings/audio/output you should be able to see sound flower 2ch and sound flower 64ch

**Comment by ManuelAbril** [comment]
Indeed, I saw them and I can selected them. However, I cannot listen any sound.

**Comment by worldsfinestdj** [comment]
That's why in the video I said to create a multi channel output by going to applications/utilities/audio midi setup and you check both built in audio and one of the sound flower options and then when you want to use sound flower you go to system settings/audio and choose the option of the multi channel device you had created.

**Comment by ghost** [comment]
Thanks mattingalls. This version is working. Just put the Soundflowerbed older app in Applications folder.

**Comment by mattingalls** [comment]
Yes, Audio MIDI Setup is your friend!  For basic tapping of Soundflower you can create an aggregate device inside there and set the aggregate to your system defaults.  FYI I have a new Soundflowerbed app in the works that will make all of this easier, but the current code base needs to be retired.

**Comment by ManuelAbril** [comment]
Thank you so much but my problem is still there. I try to explain deeper. I have an screen with integrated speakers. It is connected via HDMI to my mac mini. I have sound there but I cannot controle it in the mac itself. That is, volume up and down. On Yosemite, I installed soundflower and I changed output sound from HDMI to 2ch or 64ch and... magic, I can control the volume. However, once I did it on Capitan, I cannot. I hope you understood the issue and figure out a solution. Thanks!

**Comment by mattingalls** [comment]
@ManuelAbril this still should be possible.  you can look at the sliders in Audio MIDI Setup to confirm.
BUT i have to warn you, that volume control & mute of soundflower is BAD and it is going to be removed.  this causes WAY too many problems where users are thinking it isn't working and the volume is turned down.  In addition, it doesn't work for many cases where you want full volume going in and out but want to adjust the volume of your monitoring signal only.  The new Soundflowerbed should be able to handle all these cases, including yours..

**Comment by ManuelAbril** [comment]
So best path is to wait for your new version :)

**Comment by mattingalls** [comment]
did you try looking at Audio MIDI Setup?

**Comment by queengooborg** [comment]
@mattingalls Thanks for the link to [your version of Soundflower.](https://github.com/mattingalls/Soundflower/releases/tag/2.0b2)  It's already helped me a great amount, allowing me to stream my computer audio with [Open Broadcaster Software](https://obsproject.com/), one of the main reasons I've wanted to use Soundflower.  It should be the official version, honestly!

**Comment by lingenfr** [comment]
@mattingalls. Matt, first thanks for your work. I've used SF and the audiomixer app for years on my MBA. I got a new MB and HDMI monitor (since my TB displays are useless). I can't control the volume on the Display from the menu and had hoped to use SF. Others said it worked, but no joy for me. I am on El Capitan. I believe I installed SF (your version above) correctly and I have two SF devices in my output list. When I select either of them, I get no sound. I tried @worldsfinestvideo's method. Obviously, that doesn't work for what I am trying to do. Am I out of luck or doing something wrong. I don't have a Soundflower folder/icon in my Applications folder and no Soundflowerbed icon on my task bar. Thanks.

**Comment by beorn** [comment]
The pre-release worked for me.

**Comment by sevmorris** [comment]
@LucaTNT I'm hoping to have a multi-channel pair mod such as yours. I tried using your manual instructions, changing the path to remove /System, but this failed in Terminal. Do you have any ideas?

**Comment by nachoiz** [comment]
I have the same problem @ManuelAbril. Did you fix it? Thanks :)

**Comment by sevmorris** [comment]
Can anyone point me to an alternative app to @LucaTNT's Soundflower mod providing a series on I/O pairs?

**Comment by eriblu** [comment]
@lingenfr I think I'm doing the same thing as you.  I can confirm that the old soundflowerbed app works with the 2.0b2 Soundflower.  Here's what I did:
1. Run the 2.0b2 uninstall script, restart.
2. Custom install the Soundflowerbed app only with 1.6.6b: https://code.google.com/p/soundflower/downloads/detail?name=Soundflower-1.6.6b.dmg&can=2&q=
3. Install the 2.0b2 pkg, restart.

Soundflowerbed works and I can control system audio volume with Sunflower set to HDMI output.

**Comment by salsacurry** [comment]
Thank you, thank you, thank you, thank you!!
Thank you @mattingalls for the updated version!!!!!
&
Thank you @worldsfinestdj for explaining the thing to a complete computernerd in your video https://youtu.be/kyS7UG3wmb0

Finally after a whole day trying several things, it works!! I don't use OBS, but Screenflow and Quicktime Player it works great in both!!! 
So THANK YOU again!!!

**Comment by lingenfr** [comment]
@blufiresw thanks for the help. I was able to get SFB installed although I still can't get any sound out of my HDMI monitor. I also have an Airtame that uses SF, so I am trying to get that working first. As an immediate solution, I bought some inexpensive speakers with a volume knob (the brute force approach). Once the Airtame is working I am going to come back to this to see if I can make everything work. Cheers.

**Comment by eriblu** [comment]
@lingenfr Ah, you may have tried already, but after reinstalling the 2.0b2 pkg and restarting, are you able to select "HDMI" from SFB?

**Comment by staygritty** [comment]
I tried the link but the installation failed.  I ran the old uninstall script and got a "script error" message.  Any suggestions?

**Comment by photovirus** [comment]
All right, now I've got a new bug with Soundflower. :disappointed: 

I have a 10.11.2 beta 3 on a Macbook Air'11, and it won't boot with _signed_ Soundflower.kext installed. If I install SF without rebooting it works flawlessly... unless I reboot. Then Mac shows me a crossed circle and that's all. If I boot in verbose mode, there are no meaningful warnings.

The solution is rather simple, as with all unsigned kexts (mine was signed, but whatever, it worked):
1. Cmd+R to enter recovery mode (it works fine).
2. Fire up Terminal.
3. Remove the kext.

```
# rm -rf /Volumes/Macintosh HD/Library/Extentions/Soundflower.kext
```

And rebuild the kext cache (maybe it's optional, but I had to do it once).

```
# kextcache -u /Volumes/Macintosh HD/
```

Though, no Soundflower for me. :cry: Any ideas?

**Comment by jfcostta** [comment]
@DopeyRunr, have you found any solution for your problem? I'm also using @LucaTNT scheme for multi-track recording and now that hack in Info.plist doesn't work anymore. Any help is very appreciated.

**Comment by DopeyRunr** [comment]
@jfcostta I am using @mattingalls version successfully after turning off SIP. I'd prefer a solution that would enable me to turn SIP back on, but for now it's working OK.

**Comment by jfcostta** [comment]
I'm sorry, @DopeyRunr, but what is SIP? How do I turn it off? Thanks.

**Comment by DopeyRunr** [comment]
@jfcostta https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=mac%20osx%20disable%20sip

**Comment by mattingalls** [comment]
i don't think you have to turn anything off to use the 2.0 signed version ??

**Comment by DopeyRunr** [comment]
@mattingalls Do you mean you enabled the ability to edit the info.plist file to add more channels (see my post from 10/2)? Because that's where I was running into trouble - launching Soundflower with the edited info.plist.

**Comment by jfcostta** [comment]
That's exactly my request, I've tried to edit Info.plist in the signed version but additional channels didn't show up in Midi Setup.

**Comment by DopeyRunr** [comment]
@jfcostta Right. That's why I had to disable SIP, to allow Soundflower to load the edited info.plist. But if @mattingalls is saying there's different way to do it without disabling SIP, I'm all ears!

**Comment by jfcostta** [comment]
Yes, @DopeyRunr, same here. Awaiting @mattingalls enlightment :).

**Comment by mattingalls** [comment]
the 2.0b soundflower version installs in /Library, NOT /System/Library, so AFAIU this doesn't involve SIP

**Comment by thescottyq** [comment]
@mattingalls & @dopeyrunr

I too am trying to modify for multiple 2Ch tracks for podcasting and can't as its a signed kext. A there a way around this please?

------------------------QUOTE---------------

Thanks so much for your version of Soundflower. It's working great. The only problem I'm having is that in the past I have modified the Info.plist file to add 12 more two-channel audio devices (sample attached below). I just cut and pasted that section, changing the bolded name of the channel each time to Soundflower - A, Soundflower - B etc. 
screen shot 2015-10-02 at 8 09 10 am
However, I am unable to do as I have in the past, which is to copy this Info.plist into the package, run these commands, and use all of those "virtual audio cables."

sudo chown root:wheel /Library/Extensions/Soundflower.kext/Contents/Info.plist
sudo chmod 755 /Library/Extensions/Soundflower.kext/Contents/Info.plist
sudo kextload /Library/Extensions/Soundflower.kext

I keep getting an error on the last step:

/Library/Extensions/Soundflower.kext failed to load - (libkern/kext) not loadable (reason unspecified); check the system/kernel logs for errors or try kextutil(8).

Any ideas on how to resolve this?

In case you're curious, I use all of the additional 2 channel devices like virtual audio cables to route output from several concurrent instances of Skype into separate tracks of Adobe Audition, create a sort of mix-minus and send combined audio from the other Skypes back to each of those Skype instances. Having all of those extra 2 channel devices means I can easily edit one participant's audio without changing other participants' audio.

Thanks so much for your help!

---

**Comment by RalphWick** [comment]
Hi to All,

I followed this video on youtube : https://www.youtube.com/watch?v=rQlarrMB_u8 , and soundflower started working for me on Mac OS X - EL Capitan, try it out guys.

**Comment by RalphWick** [comment]
Try it out : https://www.youtube.com/watch?v=rQlarrMB_u8, It Started working for me on EL Capitan

**Comment by jonmrod** [comment]
I got a solution for people trying to have volume control while using two or more output devices:
Keep the soundflowerbed app outside the intial folder, run the unistall script and install 2.0b. Works great for me. (OSX El Capitan)

**Comment by keefla** [comment]
Hey everyone, so if there are any people on this thread that lack that computer whizz gene, I thought I'd share as it hope it might help out someone or at least make you laugh. I too recently downloaded el capitan for my mac and as a result lost my soundflower which I frequently utilized. I was following the directions as several of you posted, downloaded the updated soundflower, ran the uninstall Soundflower.scpt, rebooted and finally installed the soundflower.pkg. Yet, nothing kept happening and after scrolling through the mass posts with others having trouble getting this to work I assumed I was simply downloading it wrong. I thought that the soundflower would show up as an application, so I kept searching for it in my finder application folder. Finally, after three attempts to uninstall reboot and re-download, it clicked! This is not an application as I first assumed. When I finally figured this out, I checked my system preferences sounds and sure enough there it was soundflower both in my input and output settings. 
#real people problems#

**Comment by jfcostta** [comment]
@mattingalls & @DopeyRunr I've disabled SIP and edited Info.plist but additional channels didn't appear to me. I even reinstalled previous 1.6 version and got the same behavior. In both cases I rebuilt kext cache but still nothing happened. Any clue?

**Comment by DopeyRunr** [comment]
I have Soundflower.kext (with the edited Info.plist) installed in /System/Library/Extensions - maybe yours is installed in the default /Library/Extensions? Try uninstalling, rebooting, reinstalling, rebooting and see if that fixes it?

**Comment by jfcostta** [comment]
Which version do you have? I've tried both 1.6.6 and 2.0b2 (by default 2.0 installs in /Library/Extensions) and none of them showed me more than the 2ch/64ch regular channels. Right now SIP is disabled too.

**Comment by DopeyRunr** [comment]
2.0b2

**Comment by jfcostta** [comment]
Guys, thank you all, it finally worked!! Here are the steps I took:
1. Uninstall all previous versions;
2. Install SoundFlower 2.0b2 (by default it'll be copied to /Library/Extensions);
3. Edit Info.plist and add all channels you need;
4. Move SoundFlower.kext to /System/Library/Extensions (via _sudo _ to preserve permissions);
5. Reboot;
6. Do your mumbo jumbo (I used @LucaTNT trick) and enjoy multi-channel play/recording. 

At first try I disabled SIP but after all worked OK I set it back to enabled and kext still loaded fine, so it appears SIP doesn't affect this beta version of SF.

**Comment by jfcostta** [comment]
OK, a little correction: SIP is in fact required for SoundFlower work with an edited Info.plist. I had to disable kext loading in Recovery mode and after this SF loaded fine, bringing back all the missing channels.

**Comment by ashzade** [comment]
@jfcostta Can you please provide a bit more detail? I've done your 1-6 steps but it didn't work for me. Is there anything I need to besides that?

Thanks!

**Comment by jfcostta** [comment]
@ashzadeh, these 6 steps were exactly what I did, in that very strict order. But, note that loading modified kexts (this includes Info.plist) in /System/Library/Extensions requires disabling SIP, otherwise only the two default channels will be added to OSX Audio Interface. Look here for more info:
http://apple.stackexchange.com/questions/208478/how-do-i-disable-system-integrity-protection-sip-aka-rootless-on-os-x-10-11 (see 2nd answer, disabling only kext verify)

Luca's TNT trick to add multi channels can be found here:
https://lucatnt.com/2012/07/record-multi-track-skype-group-call-on-os-x/

There's one little utility to remove all previous Soundflower instances, if you need. Google is your friend. Hope that helps you.

**Comment by ashzade** [comment]
@jfcostta Thanks for the quick reply. Disabling SIP did the trick! Thank you.

**Comment by Valuecreator** [comment]
Thank you all for your comments. I've learned a great deal from you. I've since stumbled upon a 3 min YouTube video from AK Media titled "Install Soundflower on El Capitan (Easy Way)".  Here's the link www.youtube.com/watch?v=rQlarrMB_u8   

For a total newbie like me this was a blessing in disguise and sooo easy to follow. Would highly recommend you view it or, if you're like I was and you're still stumbling about - stumble no more!
